tests/e2e-evm/test/abi_basic.test.ts

```diff
@@ -0,0 +1,162 @@
+
+    // Test receive + fallback scenarios
+    { interface: "NoopReceiveNoFallback", mock: "NoopReceiveNoFallbackMock" },
+    { interface: "NoopReceivePayableFallback", mock: "NoopReceivePayableFallbackMock" },
+    { interface: "NoopReceiveNonpayableFallback", mock: "NoopReceiveNonpayableFallbackMock" },
+
+    // Test no receive + fallback scenarios
+    { interface: "NoopNoReceivePayableFallback", mock: "NoopNoReceivePayableFallbackMock" },
+    { interface: "NoopNoReceiveNonpayableFallback", mock: "NoopNoReceiveNonpayableFallbackMock" },
+  ] as ContractTestCase[];
+
+  //
+  // Client + Wallet Setup
+  //
+  let publicClient: PublicClient;
+  let walletClient: WalletClient;
+  before("setup clients", async function () {
+    publicClient = await hre.viem.getPublicClient();
+    walletClient = await hre.viem.getWalletClient(whaleAddress);
+  });
+
+  //
+  // Test each function defined in the interface ABI
+  //
+  // Only payable functions may be sent value
+  // Any function (payable, non-payable, view, pure) can be called via a transaction
+  // All functions can be provided calldata regardless of their arguments
+  for (const tc of testCases) {
+    describe(tc.interface, function () {
+      const abi = hre.artifacts.readArtifactSync(tc.interface).abi;
+      const receiveFunction = getAbiReceiveFunction(abi);
+      const fallbackFunction = getAbiFallbackFunction(abi);
+
+      let mockAddress: Address;
+      before("deploy mock", async function () {
+        mockAddress = (await hre.viem.deployContract(tc.mock)).address;
+      });
+
+      describe("State", function () {
+        it("has code set", async function () {
+          const code = await publicClient.getCode({ address: mockAddress });
+          expect(code).to.not.equal(0);
+        });
+
+        it("has nonce default of 1", async function () {
+          const nonce = await publicClient.getTransactionCount({ address: mockAddress });
+          expect(nonce).to.equal(1);
+        });
+
+        it("has a starting balance of 0", async function () {
+          const balance = await publicClient.getBalance({ address: mockAddress });
+          expect(balance).to.equal(0n);
+        });
+      });
+
+      for (const funcDesc of abi) {
+        if (funcDesc.type !== "function") {
+          continue;
+        }
+
+        const funcSelector = toFunctionSelector(toFunctionSignature(funcDesc));
+
+        describe(`ABI function ${funcDesc.name} ${funcDesc.stateMutability}`, function () {
+          const isPayable = funcDesc.stateMutability === "payable";
+
+          it("can be called", async function () {
+            const txData = { to: mockAddress, data: funcSelector, gas: 25000n };
+
+            await expect(publicClient.call(txData)).to.be.fulfilled;
+
+            const txHash = await walletClient.sendTransaction(txData);
+            const txReceipt = await publicClient.waitForTransactionReceipt({ hash: txHash });
```

Comment:
oh my, this is so much nicer than doing the same in golang 😂

Reply:
+1 I do enjoy all the built-in functionality with viem/async/await.  The API's feel much better, though golang would be a bit easier to parallelize suites like this.